### PR TITLE
Delete Kafka resources when spec.channel.enabled or spec.source.enabled is changed to false

### PR DESCRIPTION
**Pull request by aliok**
## Changes
With this PR, when the components are disabled in spec their resources are deleted.

I also did some refactoring to reuse stage execution.

------------

As written in the umbrella ticket,  https://github.com/openshift-knative/serverless-operator/issues/486, there's still a lot to do.

Verification:

1. Install eventing

2. Enable the controller by uncommenting the stuff in `knative-operator/pkg/controller/add_knativekafka.go`

3. Need to manually create the CRD since it is not in the CSV yet
```
kubectl apply -f knative-operator/deploy/crds/operator_v1alpha1_knativekafka_crd.yaml
```

4. Install:
```
cat <<EOS |kubectl apply -f -
---
apiVersion: operator.serverless.openshift.io/v1alpha1
kind: KnativeKafka
metadata:
  name: example
  namespace: knative-eventing
spec:
  source:
    enabled: true
  channel:
    enabled: true
    bootstrapServers: foo.example.com:1234
...
EOS
```

5. Watch the status of `KnativeKafka` CR, should be fine.
```
kubectl get knativekafka -n knative-eventing -o json | jq -r '.items[] | .metadata.name + ":" + (.status.conditions[] | .type + ":" + .status)'
```

6.Once status is ready, disable channel/source:
```
cat <<EOS |kubectl apply -f -
---
apiVersion: operator.serverless.openshift.io/v1alpha1
kind: KnativeKafka
metadata:
  name: example
  namespace: knative-eventing
spec:
  source:
    enabled: false
  channel:
    enabled: true
    bootstrapServers: foo.example.com:1234
...
EOS
```

7. See the resources gone and also check the status of the CR

8. Clean up:
```
kubectl delete knativekafkas example -n knative-eventing
```

### CODE REVIEW

**Comment by aliok** [comment]
The logic is like this:
- In case of CR delete, build the manifest with `BuildAll` option, and delete everyting in both KafkaChannel and KafkaSource manifests
- In other case (CR not deleted)
  - Step1: Collect resources to apply by building the manifest with `BuildEnabledOnly`. This will include resources for example for KafkaChannel if `spec.channel.enabled=true`. Then apply these resources.
  - Step2: Collect resources to delete by building the manifest with `BuildDisabledOnly`. This will include the resources for the disabled components. Later, we ensure that these resources don't exist.

I am open to any better ideas which provide these:
- In case of CR deletion, we should delete all resources in both manifests
- In case of `spec.channel.enabled` changed from `true` to `false`, we need to remove the KafkaChannel resources
- In case of `spec.channel.enabled` changed from `false` to `true`, we need to create the KafkaChannel resources

Also did a refactoring in the 2nd commit to reuse stage execution.

cc @markusthoemmes

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go
Comment (markusthoemmes) [inline knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:331]: No need to export these.

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go
Comment (markusthoemmes) [inline knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:337]: At this point, I think we should generalize this like so:

```golang
comps := map[mf.Manifest]bool{
  r.rawKafkaChannelManifest: instance.Spec.Channel.Enabled,
  r.rawKafkaSourceManifest: instance.Spec.Source.Enabled,
}

for manifest, enabled := range comps {
  	if build == ManifestBuildAll || (build == ManifestBuildEnabledOnly && enabled) || (build == ManifestBuildDisabledOnly && !enabled) {
		resources = append(resources, manifestResources()...)
	}
}
```

Comment (aliok) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:337]: yeah, why not. not a huge gain for 2 components, but sounds good.

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go
Comment (markusthoemmes) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:331]: Not a huge fan of having such a switch on a function btw, that's usually a big old cold-smell (just like boolean switches are).

I can't think of an easy way to not do that without introducing quite a bit of boilerplate, so maybe that's just fine.

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go
Comment (aliok) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:337]: hmm, can't use `mf.Manifest` as map key. It doesn't have `==` and `!=`.

can convert the map into a slice of tuples...

any alternative thoughts?

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go
Comment (aliok) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:331]: yeah, I don't like this too. I feel you. 

> I can't think of an easy way to not do that without introducing quite a bit of boilerplate, so maybe that's just fine.

Same here

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go
Comment (markusthoemmes) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go]:
```golang
appendIfNecessary := func(enabled bool, manifest mf.Manifest) {
	if build == manifestBuildAll || (build == manifestBuildEnabledOnly && enabled) || (build == manifestBuildDisabledOnly && !enabled) {
		resources = append(resources, manifest.Resources()...)
	}
}

appendIfNecessary(instance.Spec.Channel.Enabled, r.rawKafkaChannelManifest)
appendIfNecessary(instance.Spec.Source.Enabled, r.rawKafkaSourceManifest)
```

Up to you if you like that better though. I mostly wouldn't want to repeat the big `if` condition.

Comment (aliok) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go]: IMO, 2 explicit `if`s are easier to read than this. if we have a 3rd component  in the future we can convert it.

thanks for the reviews. if we're good, now let's wait until the other PR is merged.

**Comment by aliok** [comment]
@markusthoemmes @matzew 
This is rebased now and ready for your `lgtm`

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go
Comment (matzew) [inline knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:341]: that reads like WOW :-)   and is very similar to the above... :grimacing:

**Review by matzew — APPROVED** [review]
/lgtm

**Review by markusthoemmes — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
